packages/web3-shared/src/hooks/useERC20TokenDetailed.ts

```diff
@@ -1,66 +1,107 @@
+        loading: asyncToken.loading,
+        error: asyncToken.error ?? null,
+        value: tokenDetailed,
+    } as AsyncStateRetry<ERC20TokenDetailed>
+}
+
+export function useERC20TokensDetailed(listOfToken: Pick<ERC20Token, 'type' | 'address'>[]) {
+    const chainId = useChainId()
+    const listOfAddress = listOfToken.map((t) => t.address)
+    const erc20TokenContracts = useERC20TokenContracts(listOfAddress)
+    const erc20TokenBytes32Contracts = useERC20TokenBytes32Contracts(listOfAddress)
+    const initTokens = listOfToken.map((token) => ({
+        name: '',
+        address: token.address,
+        nameBytes32: '',
+        symbol: '',
+        symbolBytes32: '',
+        decimals: '0',
+    }))
+    const { value: tokens = initTokens, ...asyncTokens } = useAsyncRetry(async () => {
+        return await Promise.all(
+            listOfToken.map(async (token, i) => {
+                const erc20TokenContract = erc20TokenContracts[i]
+                const erc20TokenBytes32Contract = erc20TokenBytes32Contracts[i]
+                const _token = await getToken(erc20TokenContract, erc20TokenBytes32Contract, token)
+                return { address: token.address, ..._token }
+            }),
+        )
+    }, [chainId, erc20TokenContracts, erc20TokenBytes32Contracts])
 
-    // decimals
-    const { value: tokenDecimals = token?.decimals ?? '0', ...asyncTokenDecimals } = useAsyncRetry(
-        async () => token?.decimals ?? (await (erc20TokenContract?.methods.decimals().call() ?? '0')),
-        [erc20TokenContract, chainId],
+    const tokensDetailed = useMemo(
+        () => tokens.map((token) => (token ? createTokenDetailed(token.address, token, chainId) : null)),
+        [tokens, chainId],
     )
 
-    // token detailed
-    const tokenDetailed = useMemo(() => {
-        return {
-            type: EthereumTokenType.ERC20,
-            address: formatEthereumAddress(address),
-            chainId,
-            symbol: parseStringOrBytes32(tokenSymbol, tokenSymbolBytes32, 'UNKNOWN'),
-            name: parseStringOrBytes32(tokenName, tokenNameBytes32, 'Unknown Token'),
-            decimals: typeof tokenDecimals === 'string' ? Number.parseInt(tokenDecimals, 10) : tokenDecimals,
-        } as ERC20TokenDetailed
-    }, [tokenName, tokenNameBytes32, tokenSymbol, tokenSymbolBytes32, tokenDecimals, chainId])
+    return {
+        loading: asyncTokens.loading,
+        error: asyncTokens.error ?? null,
+        value: tokensDetailed,
+    } as AsyncStateRetry<FungibleTokenDetailed[]>
+}
 
-    const asyncList = [
-        asyncTokenName,
-        asyncTokenNameBytes32,
-        asyncTokenSymbol,
-        asyncTokenSymbolBytes32,
-        asyncTokenDecimals,
-    ]
+async function getToken(
+    erc20TokenContract: ERC20 | null,
+    erc20TokenBytes32Contract: ERC20Bytes32 | null,
+    token: Partial<ERC20TokenDetailed> | undefined,
+) {
+    const name = token?.name ?? (await (erc20TokenContract?.methods.name().call() ?? ''))
```

Comment:
Refactor this function with `Promise.allSettled` to get benefits:

+ To batch-request many promises.
+ To recover from the single point of failure.

Reply:
done

Reply:
It's still a sequential promise invoking.

Reply:
done